### PR TITLE
TLS

### PR DESCRIPTION
**Pull request by kkdawkins**
Server TLS work. License renewal coming in next PR.

### CODE REVIEW

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    pub fn get_certificate(&self) -> Result<String, Error> {
+        // First, try to directly read
+        {
+            let cert = self
+                .certificate
+                .read()
+                .expect("Locking the certificate for read failed.")
+                .to_string();
+
+            if cert.len() > 0 {
```
Comment (arsing) [inline edgelet/edgelet-http/src/certificate_manager.rs]: `if !cert.is_empty() {`

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+use failure::ResultExt;
+
+pub use crate::error::{Error, ErrorKind};
+
+const IOTEDGED_VALIDITY: u64 = 7_776_000; // 90 days
+const IOTEDGED_TLS_COMMONNAME: &str = "iotedge tls";
+
+#[derive(Clone)]
+pub struct CertificateManager<C: CreateCertificate + Clone> {
+    certificate: Arc<RwLock<String>>,
```
Comment (arsing) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Instead of treating `""` as the sentinel "non-existent" value, seems like this ought to be `Arc<RwLock<Option<String>>>` ?

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        }
+
+        // No valid cert so must create
+        let mut cert = self
+            .certificate
+            .write()
+            .expect("Locking the certificate for write failed.");
+
+        // Check that another thread hasn't already created one for us
+        if cert.to_string().len() > 0 {
```
Comment (arsing) [inline edgelet/edgelet-http/src/certificate_manager.rs]: `!cert.is_empty()`

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    pub fn get_certificate(&self) -> Result<String, Error> {
+        // First, try to directly read
+        {
+            let cert = self
+                .certificate
+                .read()
+                .expect("Locking the certificate for read failed.")
+                .to_string();
```
Comment (arsing) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Do the `.to_string()` inside the `if`. No point wasting cycles converting an empty `&str` to an empty `String` and then throwing it away.

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            .pem()
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let cert_str = String::from_utf8(cert_pem.as_ref().to_vec())
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        Ok(cert_str)
+    }
+
+    #[allow(dead_code)]
```
Comment (arsing) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Mark it `#[cfg(test)]` instead.

File: edgelet/edgelet-http/src/util/mod.rs [hunk 6]
```diff
@@ -52,32 +57,59 @@ impl StreamSelector {
...
+                Err(e) => return Err(io::Error::new(ErrorKind::Other, e)),
+            },
+            StreamSelector::TlsConnected(stream) => return stream.read(buf),
             #[cfg(windows)]
-            StreamSelector::Pipe(ref mut stream) => stream.read(buf),
-            StreamSelector::Unix(ref mut stream) => stream.read(buf),
-        }
+            StreamSelector::Pipe(stream) => return stream.read(buf),
+            StreamSelector::Unix(stream) => return stream.read(buf),
+        };
```
Comment (arsing) [inline edgelet/edgelet-http/src/util/mod.rs]: Instead of `;` and explicit `return`s, just remove both.

Same for the other functions.

File: edgelet/iotedged/Cargo.toml [hunk 7]
```diff
@@ -18,7 +18,8 @@ sha2 = "0.7.0"
 tokio = "0.1.8"
 tokio-signal = "0.2"
 url = "1.7"
-
+url_serde = "0.2"
+native-tls = "0.2"
 hsm = { path = "../hsm-rs"}
```
Comment (arsing) [inline edgelet/iotedged/Cargo.toml]: Keep the blank line that separates crates.io deps from workspace deps.

File: edgelet/iotedged/src/error.rs [hunk 8]
```diff
@@ -79,6 +79,7 @@ impl From<Context<ErrorKind>> for Error {
 pub enum InitializeErrorReason {
     CreateMasterEncryptionKey,
     CreateSettingsDirectory,
+    CreateTLSCertificate,
```
Comment (arsing) [inline edgelet/iotedged/src/error.rs]: `CreateTlsCertificate`

File: edgelet/iotedged/src/lib.rs [hunk 9]
```diff
@@ -881,6 +889,8 @@ where
     let label = "mgmt".to_string();
     let url = settings.listen().management_uri().clone();
 
+    let cert_manager = cert_manager.clone();
```
Comment (arsing) [inline edgelet/iotedged/src/lib.rs]: If you want to own it here, why clone a borrow instead of taking it by value in the first place?

Same in `start_workload`

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 3 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: I am holding a write lock here, so I don't think I can call into a function that tries to open a read lock, right?

File: edgelet/iotedged/Cargo.toml (hunk 7 above)
Comment (arsing) [reply edgelet/iotedged/Cargo.toml]: The blank line separates crates.io deps from workspace deps. ie it goes before `hsm`.

Also, keep the deps in alphabetical order, so `native-tls` goes before `serde_json`.

File: edgelet/iotedged/src/lib.rs [hunk 10]
```diff
@@ -515,14 +517,27 @@ where
...
+        Arc::clone(&cert_manager),
+    );
 
     let workload = start_workload(
         &settings,
         key_store,
         &runtime,
         work_rx,
         crypto,
+        Arc::clone(&cert_manager),
```
Comment (arsing) [inline edgelet/iotedged/src/lib.rs]: Don't need to create a second clone. Just pass the original `cert_manager`.

File: edgelet/iotedged/src/lib.rs [hunk 11]
```diff
@@ -881,6 +898,8 @@ where
     let label = "mgmt".to_string();
     let url = settings.listen().management_uri().clone();
 
+    let cert_manager = cert_manager.clone();
```
Comment (arsing) [inline edgelet/iotedged/src/lib.rs]: You don't need to `clone`.

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+                    .create_cert()
+                    .with_context(|_| ErrorKind::CertificateCreationError)?;
+                Ok(cert.get_or_insert(new_cert).to_string())
+            }
+        }
+    }
+
+    fn create_cert(&self) -> Result<String, Error> {
+        let edgelet_cert_props = CertificateProperties::new(
+            IOTEDGED_VALIDITY,
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Should we make this an input param to `CertificateManager` (maybe eventually sourced from `config.yaml`)?

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+            IOTEDGED_TLS_COMMONNAME.to_string(),
+            CertificateType::Server,
+            "iotedge-tls".to_string(),
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Should this be a literal string or an input? Not sure what the "alias" name in the cert is used for. Might be useful to check with Mahesh.

File: edgelet/iotedged/src/lib.rs [hunk 14]
```diff
@@ -515,14 +517,27 @@ where
     let (mgmt_tx, mgmt_rx) = oneshot::channel();
     let (work_tx, work_rx) = oneshot::channel();
 
-    let mgmt = start_management(&settings, &runtime, &id_man, mgmt_rx);
+    let cert_manager = CertificateManager::new(crypto.clone()).context(ErrorKind::Initialize(
+        InitializeErrorReason::CreateTlsCertificate,
```
Comment (avranju) [inline edgelet/iotedged/src/lib.rs]: We don't actually create a certificate till `get_certificate` is called no? If we change `CertificateManager::new` so that it no longer returns a `Result` then I suppose this error handling will go away.

File: edgelet/iotedged/src/lib.rs [hunk 15]
```diff
@@ -515,14 +517,27 @@ where
...
+    ))?;
+
+    let cert_manager = Arc::new(cert_manager);
+
+    let mgmt = start_management(
+        &settings,
+        &runtime,
+        &id_man,
+        mgmt_rx,
+        Arc::clone(&cert_manager),
```
Comment (avranju) [inline edgelet/iotedged/src/lib.rs]: Couldn't we say `cert_manager.clone()`?

File: edgelet/iotedged/src/lib.rs [hunk 16]
```diff
@@ -921,21 +941,24 @@ where
...
         + 'static,
+    CE: CreateCertificate + Clone,
     W: WorkloadConfig + Clone + Send + Sync + 'static,
 {
     info!("Starting workload API...");
 
     let label = "work".to_string();
     let url = settings.listen().workload_uri().clone();
 
+    let cert_manager = cert_manager.clone();
```
Comment (avranju) [inline edgelet/iotedged/src/lib.rs]: As Arnav pointed out above, this clone should not be necessary.

File: edgelet/edgelet-http/src/util/mod.rs [hunk 17]
```diff
@@ -52,31 +57,58 @@ impl StreamSelector {
...
+            StreamSelector::Tcp(stream) => stream.read(buf),
+            StreamSelector::TlsConnecting(stream) => match stream.poll() {
+                Ok(Async::Ready(stream)) => {
+                    *self = StreamSelector::TlsConnected(stream);
+                    self.read(buf)
+                }
+                Ok(Async::NotReady) => Err(ErrorKind::WouldBlock.into()),
+                Err(e) => Err(io::Error::new(ErrorKind::Other, e)),
+            },
+            StreamSelector::TlsConnected(stream) => stream.read(buf),
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/mod.rs]: Do we have tests for the implementation in this file? If not, should we?

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 12 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: Good thinking!

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+use std::sync::{Arc, RwLock};
+
+use edgelet_core::crypto::{Certificate, CreateCertificate};
+use edgelet_core::{CertificateIssuer, CertificateProperties, CertificateType};
+use failure::ResultExt;
+
+pub use crate::error::{Error, ErrorKind};
+
+const IOTEDGED_VALIDITY: u64 = 7_776_000; // 90 days
+const IOTEDGED_TLS_COMMONNAME: &str = "iotedge tls";
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: It is now an input as you can define your own props when you create the manager

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 19]
```diff
@@ -0,0 +1,175 @@
+use std::sync::{Arc, RwLock};
+
+use edgelet_core::crypto::{Certificate, CreateCertificate};
+use edgelet_core::CertificateProperties;
+use failure::ResultExt;
+
+pub use crate::error::{Error, ErrorKind};
+
+#[derive(Clone)]
+pub struct CertificateManager<C: CreateCertificate + Clone> {
```
Comment (darobs) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Do you need the restriction on `C` here?

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    // Convenience function since native-tls does not yet support PEM
+    // and since everything else uses PEM certificates, we want to keep
+    // the actual storage of the certificate in the PEM format.
+    #[cfg(unix)]
+    pub fn get_pkcs12_certificate(&self) -> Result<Vec<u8>, Error> {
+        let stored_cert_bundle = self
+            .get_certificate()
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Since `get_certificate` already returns `Error` this `with_context` may not be necessary.

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+
+        let cert = stored_cert_bundle.0.as_bytes();
+
+        let mut certs =
+            X509::stack_from_pem(cert).with_context(|_| ErrorKind::CertificateConverstionError)?;
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: `s/CertificateConverstionError/CertificateConversionError/g`

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+
+        let mut ca_certs = Stack::new().with_context(|_| ErrorKind::CertificateConverstionError)?;
+        for cert in certs.split_off(1) {
+            ca_certs
+                .push(cert)
+                .with_context(|_| ErrorKind::CertificateConverstionError)?;
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Seeing all these `with_context` calls, it makes me wonder if it's better to simply have a conversion that converts from openssl errors into our `Error`!

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+        }
+
+        let key = PKey::private_key_from_pem(stored_cert_bundle.1.as_bytes())
+            .expect("Issue getting private key");
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Maybe the error message should be `Error while parsing private key pem`.

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let key_str = String::from_utf8(pk_bytes.as_bytes().to_vec())
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        Ok((cert_str, key_str))
+    }
+
+    #[cfg(test)]
+    fn has_certificate(&self) -> bool {
```
Comment (avranju) [inline edgelet/edgelet-http/src/certificate_manager.rs]: Were you going to move this function into the test module below?

File: edgelet/edgelet-http/src/lib.rs [hunk 25]
```diff
@@ -186,6 +210,38 @@ impl HyperExt for Http {
                     .with_context(|_| ErrorKind::BindListener(BindListenerType::Address(addr)))?;
                 Incoming::Tcp(listener)
             }
+            #[cfg(unix)]
+            HTTPS_SCHEME => {
+                let addr = url
+                    .to_socket_addrs()
+                    .context(ErrorKind::InvalidUrl(url.to_string()))?
```
Comment (avranju) [inline edgelet/edgelet-http/src/lib.rs]: Nit: Wish there was a way to avoid the alloc when calling `url.to_string()` when there's no error.

File: edgelet/edgelet-http/src/util/incoming.rs [hunk 26]
```diff
@@ -34,6 +51,60 @@ impl Stream for Incoming {
...
             }
+            #[cfg(unix)]
+            Incoming::Tls(ref mut listener, ref mut acceptor, ref mut connections) => {
+                // check if we have a tcp connection and if we do then kick off TLS handshake
+                // and store the future representing that operation in "connections"
+                if let Err(err) = listener.poll_accept().map(|val| {
+                    if let Async::Ready((tcp_stream, addr)) = val {
+                        connections
+                            .lock()
+                            .expect("Unable to lock the connection mutex")
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/incoming.rs]: Nit: `s/connection/connections`

File: edgelet/edgelet-http/src/util/incoming.rs [hunk 27]
```diff
@@ -34,6 +51,60 @@ impl Stream for Incoming {
...
+                            .push((acceptor.accept(tcp_stream), IncomingSocketAddr::Tcp(addr)));
+                    }
+                }) {
+                    return Err(err);
+                }
+
+                // Take a read lock on our pending connection list returning a tuple representing an index
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/incoming.rs]: Nit: since this is a mutex there's no "read" lock -- just a lock.

File: edgelet/edgelet-http/src/util/incoming.rs [hunk 28]
```diff
@@ -34,6 +51,60 @@ impl Stream for Incoming {
...
+                // + the state on the poll
+                let mut connections = connections
+                    .lock()
+                    .expect("Unable to lock the connection mutex");
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/incoming.rs]: Nit: `s/connection/connections`

File: edgelet/edgelet-http/src/util/incoming.rs [hunk 29]
```diff
@@ -34,6 +51,60 @@ impl Stream for Incoming {
...
+                let val = connections
+                    .iter_mut()
+                    .map(|(fut, _)| fut.poll())
+                    .enumerate()
+                    .find(|(_, result)| match result {
+                        Ok(v) => v.is_ready(),
+                        Err(_) => true,
+                    });
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/incoming.rs]: So, lines 74-81 is basically doing the following:

>_From the list of on-going TLS handshakes, find the first one that has either completed successfully or has failed and return a tuple containing its index and state._

Maybe we should make the comment above (on line 69) a bit more descriptive to indicate this. This code looks a little bit hard to parse otherwise.

File: edgelet/edgelet-http/src/util/incoming.rs [hunk 30]
```diff
@@ -34,6 +51,60 @@ impl Stream for Incoming {
...
+                                    Async::Ready(Some((StreamSelector::Tls(tls_stream), addr)))
+                                }
+                                #[cfg(windows)]
+                                {
+                                    Async::Ready(Some(
+                                        (Box::new(StreamSelector::Tls(tls_stream), addr)),
+                                    ))
+                                }
+                            }
+                            Ok(_) => unreachable!(),
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/incoming.rs]: Maybe we should add a comment describing why this alternative should be unreachable. I suppose it's because the predicate passed to the `find` call on line 78 explicitly filters for the `is_ready` state.

File: edgelet/edgelet-http/src/util/incoming.rs [hunk 31]
```diff
@@ -34,6 +51,60 @@ impl Stream for Incoming {
...
+                match val {
+                    Some((i, result)) => {
+                        let (_, addr) = connections.remove(i);
+                        match result {
+                            Ok(Async::Ready(tls_stream)) => {
+                                #[cfg(not(windows))]
+                                {
+                                    Async::Ready(Some((StreamSelector::Tls(tls_stream), addr)))
+                                }
+                                #[cfg(windows)]
```
Comment (avranju) [inline edgelet/edgelet-http/src/util/incoming.rs]: Since the enum variant `Incoming::Tls` is not available on Windows, I am wondering why this code compiles on Windows causing the "variant too large" clippy error. Clippy doesn't care for `cfg` annotations?

File: edgelet/edgelet-http/tests/tls.rs [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+    let manager = CertificateManager::new(crypto.clone(), edgelet_cert_props);
+
+    let recognizer = RegexRoutesBuilder::default()
+        .get(Version::Version2018_06_28, "/route1/hello", route1)
+        .finish();
+    let router = Router::from(recognizer);
+
+    Http::new()
+        .bind_url(
+            Url::parse("https://127.0.0.1:8881").unwrap(),
```
Comment (avranju) [inline edgelet/edgelet-http/tests/tls.rs]: Should we use an arbitrary unused TCP port instead of `8881`? See `edgelet_test_utils::get_unused_tcp_port`.

File: edgelet/edgelet-http/examples/identities.rs [hunk 33]
```diff
@@ -83,7 +84,10 @@ fn main() {
     let addr = "tcp://0.0.0.0:8080".parse().unwrap();
 
     println!("Starting server on {}", addr);
-    let run = Http::new().bind_url(addr, router).unwrap().run();
+    let run = Http::new()
+        .bind_url(addr, router, None::<&CertificateManager<Crypto>>)
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edgelet/edgelet-http/examples/identities.rs]: I did find that amusing - it is like we are taunting the caller by saying "We are returning nothing... but know that you are missing out on a CertificateManager" :)

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 34]
```diff
@@ -0,0 +1,243 @@
...
+            .certificate
+            .write()
+            .expect("Locking the certificate for write failed.");
+
+        if let Some(cert) = cert.as_ref() {
+            Ok(cert.clone())
+        } else {
+            let (new_cert, private_key) = self
+                .create_cert()
+                .with_context(|_| ErrorKind::CertificateCreationError)?;
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: Well, the issue here is that I then need to insert the tuple value into the option using get_or_insert. However, your suggestion above to use a Certificate struct makes me want to re-think this a bit.

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 20 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: See above.

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 23 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: That is much more descriptive - however, I did want to discuss how I can tell if an error is bubbled up to the user vs just put in our logs. Sometimes, I would like to provide more information but I am unsure if it's appropriate to show the user so much of our inner workings.

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 24 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: Will follow up on this - I use this function to tell if my get_certificate code is working since if no certificate exists, it will create one. It operates on the Cert Manager struct and AFAIK, it must be where it is (unless I am missing some rust test code magic :) ) .

File: edgelet/edgelet-http/src/util/incoming.rs (hunk 31 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/util/incoming.rs]: Well the StreamSelector was not using cfg(Unix) for the TLS stream since none of that code relied on OpenSSL. I suppose that we could use cfg on *all* TLS code, and then this would not happen on Windows.

File: edgelet/edgelet-http/src/certificate_manager.rs [hunk 35]
```diff
@@ -0,0 +1,243 @@
...
+            .get_private_key()
+            .with_context(|_| ErrorKind::CertificateCreationError)?;
+
+        let pk = match cert_private_key {
+            Some(pk) => pk,
+            None => panic!("did not expect reference private key"),
+        };
+
+        let pk_bytes = match pk {
+            PrivateKey::Ref(_) => panic!("did not expect reference private key"),
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: Spoke with Mahesh offline and made a comment about reasoning

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 22 above)
Comment (kkdawkins) [reply edgelet/edgelet-http/src/certificate_manager.rs]: Spoke with Arnav offline. Context is preferred so that when we "pretty print" errors, we have a chain "caused by ... " for the user.

File: edgelet/edgelet-http/src/certificate_manager.rs (hunk 24 above)
Comment (avranju) [reply edgelet/edgelet-http/src/certificate_manager.rs]: I see. That's fine then.

File: edgelet/edgelet-http/src/lib.rs (hunk 25 above)
Comment (avranju) [reply edgelet/edgelet-http/src/lib.rs]: AFAICT there really isn't a way to do this at this time.

File: edgelet/edgelet-http/src/util/incoming.rs (hunk 31 above)
Comment (avranju) [reply edgelet/edgelet-http/src/util/incoming.rs]: You're right, the problem was on the `StreamSelector` and not `Incoming`. I am OK with leaving this be.

File: edgelet/iotedged/src/lib.rs [hunk 36]
```diff
@@ -515,14 +518,25 @@ where
...
     let (work_tx, work_rx) = oneshot::channel();
 
-    let mgmt = start_management(&settings, &runtime, &id_man, mgmt_rx);
+    let edgelet_cert_props = CertificateProperties::new(
+        IOTEDGED_VALIDITY,
+        IOTEDGED_TLS_COMMONNAME.to_string(),
+        CertificateType::Server,
+        "iotedge-tls".to_string(),
+    )
+    .with_issuer(CertificateIssuer::DeviceCa);
```
Comment (avranju) [inline edgelet/iotedged/src/lib.rs]: It'll be nice to have these parameters, i.e., validity, TLS name and alias come from configuration or an env variable. That way customers can override it. But we can do that via a different PR and just file a task for this.

**Comment by avranju** [comment]
@arsing Could you approve the PR in case you have no other comments please?

File: edgelet/iotedged/src/lib.rs (hunk 36 above)
Comment (kkdawkins) [reply edgelet/iotedged/src/lib.rs]: @avranju created task # 4318495
